### DISCUSSION TITLE
Question about "bitcoincore_zmq::Message::Tx"

### DISCUSSION THREAD

**Discussion by kkthmyh** [antonilol/rust-bitcoincore-zmq#15] [Q&A] [answered]
When I monitor new transactions, what does this u32 mean? I printed it on the console and found that it is auto-incrementing, so will it grow until u32::MAX? Then wouldn't the program panic?
<img width="1090" alt="image" src="https://github.com/antonilol/rust-bitcoincore-zmq/assets/34261479/df8dfedf-d6da-4a2d-aa73-1e4a806c5355">
```rust
 for msg in bitcoincore_zmq::subscribe_receiver(&["tcp://127.0.0.1:28332"]).unwrap() {
        match msg {
            Ok(msg) => {
                match msg {
                    bitcoincore_zmq::Message::Tx(tx, seq) => {
                        println!("{:?}", seq);
                        println!("{:?}", tx.txid());
                    }
                    _ => {}
                }
            }
            Err(err) => println!("Error receiving message: {err}"),
        }
    }
```

**Comment by antonilol** [comment, marked as answer]
I wrote some documentation on that, but it is on the `sequence` method of `Message` ([see docs here](https://docs.rs/bitcoincore-zmq/latest/bitcoincore_zmq/enum.Message.html#method.sequence)).

This number is in Bitcoin Core for each zmq publisher, it is [defined as uint32_t](https://github.com/bitcoin/bitcoin/blob/master/src/zmq/zmqpublishnotifier.h#L21), and [incremented](https://github.com/bitcoin/bitcoin/blob/master/src/zmq/zmqpublishnotifier.cpp#L215) every time a zmq message is sent, without additional special logic, so after u32::MAX it rolls over back to 0.

This number can be used to track if any messages are lost. Lets say you have to restart your program (but not bitcoin core), if the program records the last seen sequence number before restarting it can compare it to the first seen one after booting up, and know if and how many messages were missed.

If you need more info on Bitcoin Core zmq specifics like this, I linked some documentation in the [readme, "Useful resources"](https://github.com/antonilol/rust-bitcoincore-zmq?tab=readme-ov-file#useful-resources).

**Comment by kkthmyh** [comment]
> I wrote some documentation on that, but it is on the `sequence` method of `Message` ([see docs here](https://docs.rs/bitcoincore-zmq/latest/bitcoincore_zmq/enum.Message.html#method.sequence)).
> 
> This number is in Bitcoin Core for each zmq publisher, it is [defined as uint32_t](https://github.com/bitcoin/bitcoin/blob/master/src/zmq/zmqpublishnotifier.h#L21), and [incremented](https://github.com/bitcoin/bitcoin/blob/master/src/zmq/zmqpublishnotifier.cpp#L215) every time a zmq message is sent, without additional special logic, so after u32::MAX it rolls over back to 0.
> 
> This number can be used to track if any messages are lost. Lets say you have to restart your program (but not bitcoin core), if the program records the last seen sequence number before restarting it can compare it to the first seen one after booting up, and know if and how many messages were missed.
> 
> If you need more info on Bitcoin Core zmq specifics like this, I linked some documentation in the [readme, "Useful resources"](https://github.com/antonilol/rust-bitcoincore-zmq?tab=readme-ov-file#useful-resources).

Thank you very much, I'll read the documentation.
